src/modm/ui/color/rgb.hpp

```diff
@@ -24,102 +19,210 @@
+	constexpr RgbD(const C& other)
+		: red_(other.red()), green_(other.green()), blue_(other.blue())
 	{}
 
-	/**
-	 * Convertion Constructor for HSV Color
-	 *
-	 * @param hsv	HSV Color
-	 */
-	template<std::unsigned_integral U>
-	constexpr RgbT(const HsvT<U>& hsv);
-
-	/**
-	 * Convertion Constructor for Brightness
-	 *
-	 * @param brightness	Brightness 'Color'-object
-	 */
-	// TODO Plump conversion, implement the right way
-	template<std::unsigned_integral U>
-	constexpr RgbT(const BrightnessT<U> brightness)
-		: red(brightness), green(brightness), blue(brightness)
+	template<ColorGray C>
+	constexpr RgbD(const C &gray)
+		: red_(gray), green_(gray), blue_(gray)
 	{}
 
-	/**
-	 * Convertion Constructor for RGB565 Color
-	 *
-	 * @param rgb565	RGB565 Color
-	 */
-	constexpr RgbT(const Rgb565& rgb565)
-		: red((rgb565.color >> 8) & 0xF8),
-		  green((rgb565.color >> 3) & 0xFC),
-		  blue(rgb565.color << 3)
-	{}
+	template<ColorHsv C>
+	constexpr RgbD(const C& hsv)
+	{
+		// OPTIMIZE No need to calculate sharper than the output
+		// Develop CalcType from types of conversion target: RedType, GreenType and BlueType.
+		using CalcType = C::ValueType;
+		using T = CalcType::T;
+
+		using WideType = modm::WideType<T>;
+		using WideWideType = modm::WideType<WideType>;
+		static_assert(!std::is_same_v<WideType, WideWideType>, "C::T too big");
+
+		const T hue = CalcType(hsv.hue());
+		const T saturation = CalcType(hsv.saturation());
+		const T value = CalcType(hsv.value());
+
+		const WideType vs = value * saturation;
+		const WideType h6 = 6 * hue;
+
+		T i = h6 >> CalcType::Digits;
+		WideType f = ((i | 1) << CalcType::Digits) - h6;
+		if (i & 1) f = -f;
+
+		CalcType p(((value << CalcType::Digits) - vs) >> CalcType::Digits);
+		CalcType u(((WideWideType(value) << 2 * CalcType::Digits) - WideWideType(vs) * f) >> 2 * CalcType::Digits);
+
+		switch (i)
+		{
+			case 0: red_ = hsv.value(); green_ = u; blue_ = p; break;
+			case 1: red_ = u; green_ = hsv.value(); blue_ = p; break;
+			case 2: red_ = p; green_ = hsv.value(); blue_ = u; break;
+			case 3: red_ = p; green_ = u; blue_ = hsv.value(); break;
+			case 4: red_ = u; green_ = p; blue_ = hsv.value(); break;
+			case 5: red_ = hsv.value(); green_ = p; blue_ = u; break;
+		}
+	}
+
+	// accessors
+	const RedType red() const { return red_; }
+	const GreenType green() const { return green_; }
+	const BlueType blue() const { return blue_; }
+
+	RedType& red() { return red_; }
+	GreenType& green() { return green_; }
+	BlueType& blue() { return blue_; }
+
+	// operator +=, -=, *=, /=
+	RgbD& operator+=(const RgbD& other) {
+		red_ += other.red();
+		green_ += other.green();
+		blue_ += other.blue();
+		return *this;
+	}
+
+	RgbD& operator-=(const RgbD& other) {
+		red_ -= other.red();
+		green_ -= other.green();
+		blue_ -= other.blue();
+		return *this;
+	}
+
+	RgbD& operator*=(const RgbD& other) {
+		red_ *= other.red();
+		green_ *= other.green();
+		blue_ *= other.blue();
+		return *this;
+	}
```

Comment:
Is component-wise multiplication / division of color values a sensible operation? What does this mean?

Reply:
It's a cheap ways to darken/lighten a color. Looks useful to me.

Reply:
> It's a cheap ways to darken/lighten a color. Looks useful to me.

If the data type allowed for fractional values that could work but the components are integer here.